<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цели-исследования" data-toc-modified-id="Цели-исследования-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Цели исследования</a></span></li><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Изучение данных</a></span></li><li><span><a href="#SQL-запросы" data-toc-modified-id="SQL-запросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>SQL-запросы</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года." data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года.</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку." data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку.</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры." data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры.-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками." data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками.-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Проект по SQL

## Цели исследования

Наша компания купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Изучение данных

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [4]:
query = ''' SELECT *
            FROM books
            LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Таблица `books` содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

In [5]:
query = ''' SELECT *
            FROM authors
            LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Таблица `authors` содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

In [6]:
query = ''' SELECT *
            FROM publishers
            LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Таблица `publishers` содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

In [7]:
query = ''' SELECT *
            FROM ratings
            LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Таблица `ratings` содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

In [8]:
query = ''' SELECT *
            FROM reviews
            LIMIT 5;
'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Таблица `reviews` содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

## SQL-запросы

### Посчитайте, сколько книг вышло после 1 января 2000 года.

In [9]:
query = ''' SELECT COUNT(book_id)
            FROM books
            WHERE publication_date > '2000-01-01';
'''
pd.io.sql.read_sql(query, con = engine)

count
0    819

### Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [10]:
query = ''' SELECT title,
                   COALESCE (count, 0) AS count,
                   avg
            FROM books AS b
            LEFT OUTER JOIN (SELECT book_id,
                                    COUNT(review_id)
                             FROM reviews
                             GROUP BY book_id) AS r_count ON b.book_id = r_count.book_id
            LEFT OUTER JOIN (SELECT book_id,
                                    AVG(rating)
                             FROM ratings
                             GROUP BY book_id) AS r_avg ON b.book_id = r_avg.book_id
            ORDER BY count DESC,
                     avg DESC;

'''
pd.io.sql.read_sql(query, con = engine)

title  count       avg
0                              Twilight (Twilight  #1)      7  3.662500
1    Harry Potter and the Prisoner of Azkaban (Harr...      6  4.414634
2    Harry Potter and the Chamber of Secrets (Harry...      6  4.287500
3                                       The Book Thief      6  4.264151
4                                     The Glass Castle      6  4.206897
..                                                 ...    ...       ...
995                               Leonardo's Notebooks      0  4.000000
996  Disney's Beauty and the Beast (A Little Golden...      0  4.000000
997                          Essential Tales and Poems      0  4.000000
998    Anne Rice's The Vampire Lestat: A Graphic Novel      0  3.666667
999                            The Natural Way to Draw      0  3.000000

[1000 rows x 3 columns]

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [11]:
query = ''' SELECT publisher,
                   count
            FROM publishers AS p
            JOIN (SELECT publisher_id,
                         COUNT(book_id)
                  FROM books
                  WHERE num_pages > 50
                  GROUP BY publisher_id
                  ORDER BY COUNT(book_id) DESC
                  LIMIT 1) AS best_p ON p.publisher_id = best_p.publisher_id;
'''
pd.io.sql.read_sql(query, con = engine)

publisher  count
0  Penguin Books     42

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [12]:
query = ''' SELECT a.author,
                   AVG(avg)
            FROM (SELECT book_id,
                         COUNT(rating_id),
                         AVG (rating)
                  FROM ratings 
                  GROUP BY book_id
                  HAVING COUNT(rating_id) >= 50) AS more_than_fifty
            JOIN books AS b ON more_than_fifty.book_id = b.book_id
            JOIN authors AS a ON b.author_id = a.author_id
            GROUP BY a.author
            ORDER BY AVG(avg) DESC
            LIMIT 1;
'''
pd.io.sql.read_sql(query, con = engine)

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [13]:
query = ''' SELECT AVG(count_of_reviews)
            FROM (SELECT username,
                         COUNT (review_id) AS count_of_reviews
                  FROM reviews
                  GROUP BY username) AS user_reviews
            JOIN (SELECT username,
                  COUNT (rating_id) AS count_of_ratings
                  FROM ratings
                  GROUP BY username
                  HAVING COUNT (rating_id) > 50) AS user_ratings ON user_reviews.username = user_ratings.username;
'''
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

## Выводы

- После 1 января 2000 года вышло 819 книг.
- Топ-3 книг по количеству обзоров и средней оценке:
 1. Сумерки (7 обзоров, средняя оценка 3.66)
 2. Гарри Поттер и узник Азкабана (7 обзоров, средняя оценка 4.41)
 3. Гарри Поттер и тайная комната (6 обзоров, средняя оценка 4.29).
- Наибольшее число книг выпустило издательство Penguin Books.
- Джоан Роулинг - автор с самой высокой средней оценкой книг (4.28).
- Пользователи, которые поставили больше 50 оценок, написали в среднем 24 обзора.